example_allocators/bump_allocator/bump_alloc.c

```diff
@@ -51,16 +51,23 @@ void init_alloc(int size_in_bytes)
 char *bump_alloc(int len)
 {
 	char *chunk = 0;
+	size_t rounded_len; /* for CHERI alignment */
+
 	if (count + len <= max)
 	{
 		/* bump the pointer */
 		chunk = buffer + count;
 
+		/* check we can represent the capability accurately,
+		 * see p30 of CHERI C/C++ Prog Guide */
+		chunk = __builtin_align_up(chunk, ~cheri_representable_alignment_mask(len) + 1);
+		rounded_len = cheri_representable_length(len);
+
 		/* restrict capability range before returning ptr */
-		chunk = cheri_bounds_set_exact(chunk, len);
+		chunk = cheri_bounds_set_exact(chunk, rounded_len);
```

Comment:
To be clear, this will crash if there isn't space in `chunk`'s bounds. I think that's Ok here, and in many bump-pointer applications, but a production allocator will probably want to handle it explicitly (e.g. returning NULL, like malloc).

Reply:
>To be clear, this will crash if there isn't space in chunk's bounds

Are you saying that at least one problem is that `rounded_len` will over-approximate the memory so we'll end up providing bigger bounds than we should? Or something else as well?

Reply:
oops! - I think you are right - I only check we have enough space for len bytes, not enough space for rounded_len bytes ... that's very silly of me. Will fix up now :-)

Reply:
Well, here's an interesting question: do we *want* to allocate the `rounded_len` number of bytes because that will often waste some memory, potentially mega/gigabytes. So one way one to look at the capability is that it guarantees to cover all of our heap, but possibly also a lot of things beyond the heap too. But that might still be better than not having a capability at all?

Reply:
I think I meant `buffer`, sorry.

There are (at least) three interesting cases:
1. The caller asks for more memory than is actually available, so `len > length(buffer)`.
2. `len` is large and unrepresentable so we have to round it up.
3. The base (at `buffer + count`) is insufficiently aligned for the `rounded_len`.

Case 1 is obviously an out-of-memory condition.

I don't think cases 2 and 3 can cause an error by themselves (in the absence of case 1). We pad at one end of the allocation, but I don't think that padding can put `chunk`'s limit outside `buffer` because (in the absence of 1) `len <= length(buffer)` and so it has at least as fine a precision. This is intuition at best; a proof would be useful, I think.

The combination of 2 and 3 can pad at both ends, and that can make an allocation fail even if there are enough actual bytes remaining. It'll be hard to hit (and therefore test) in practice.

Reply:
@ltratt I think that's a per-allocator policy decision. If permissions will differ (e.g. the W^X example) then you need to, to prevent overlaps that could be security issues. In other situations (especially if allocations will be further sub-divided anyway) then it may be acceptable.

The wastage may be mega/gigabytes but still a very small compared to the allocation, and much larger than hardware pages (so it's just VA space, not physical memory).

Reply:
Point taken!

Reply:
hi @jacobbramley thanks for the detailed feedback. I am now considering all three interesting cases you mention above. Previously, I was unaware of case (3) so this was a useful exercise.